### ISSUE TITLE
binary wheels can fail to install when CUDA_HOME not set: "nvcc not found"

### ISSUE THREAD

**Issue by tmm1** [Dao-AILab/flash-attention#447]
```
Building wheels for collected packages: flash-attn
  Building wheel for flash-attn (setup.py) ... error
  error: subprocess-exited-with-error

  × python setup.py bdist_wheel did not run successfully.
  │ exit code: 1
  ╰─> [20 lines of output]
      fatal: not a git repository (or any of the parent directories): .git
      running bdist_wheel
      Traceback (most recent call last):
        File "<string>", line 2, in <module>
        File "<pip-setuptools-caller>", line 34, in <module>
        File "/tmp/pip-install-56tllxed/flash-attn_e3fae8fde0f54220827a3730f9387260/setup.py", line 293, in <module>
          setup(
        File "/home/teknium/finetune-study/finetune-study/lib/python3.10/site-packages/setuptools/__init__.py", line 153, in setup
          return distutils.core.setup(**attrs)
        File "/usr/lib/python3.10/distutils/core.py", line 148, in setup
          dist.run_commands()
        File "/usr/lib/python3.10/distutils/dist.py", line 966, in run_commands
          self.run_command(cmd)
        File "/usr/lib/python3.10/distutils/dist.py", line 985, in run_command
          cmd_obj.run()
        File "/tmp/pip-install-56tllxed/flash-attn_e3fae8fde0f54220827a3730f9387260/setup.py", line 249, in run
          raise_if_cuda_home_none("flash_attn")
        File "/tmp/pip-install-56tllxed/flash-attn_e3fae8fde0f54220827a3730f9387260/setup.py", line 86, in raise_if_cuda_home_none
          raise RuntimeError(
      RuntimeError: flash_attn was requested, but nvcc was not found.  Are you sure your environment has nvcc available?  If you're installing within a container from https://hub.docker.com/r/pytorch/pytorch, only images whose names contain 'devel' will provide nvcc.
      [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  ERROR: Failed building wheel for flash-attn
  Running setup.py clean for flash-attn
Failed to build flash-attn
Installing collected packages: flash-attn
  Running setup.py install for flash-attn ... done
  DEPRECATION: flash-attn was installed using the legacy 'setup.py install' method, because a wheel could not be built for it. A possible replacement is to fix the wheel build issue reported above. Discussion can be found at https://github.com/pypa/pip/issues/8368
Successfully installed flash-attn-2.0.6
```

**Comment by tmm1** [comment]
w/ `pip -v`

```
Using pip 22.0.2 from /home/teknium/finetune-study/finetune-study/lib/python3.10/site-packages/pip (python 3.10)
Collecting flash-attn==2.0.6
  Using cached flash_attn-2.0.6.tar.gz (2.4 MB)
  Running command python setup.py egg_info
  fatal: not a git repository (or any of the parent directories): .git
  Traceback (most recent call last):
    File "<string>", line 2, in <module>
    File "<pip-setuptools-caller>", line 34, in <module>
    File "/tmp/pip-install-wub410gf/flash-attn_06d2dd1c6ec841859f0d8ad4972cc418/setup.py", line 140, in <module>
      raise_if_cuda_home_none("flash_attn")
    File "/tmp/pip-install-wub410gf/flash-attn_06d2dd1c6ec841859f0d8ad4972cc418/setup.py", line 86, in raise_if_cuda_home_none
      raise RuntimeError(
  RuntimeError: flash_attn was requested, but nvcc was not found.  Are you sure your environment has nvcc available?  If you're installing within a container from https://hub.docker.com/r/pytorch/pytorch, only images whose names contain 'devel' will provide nvcc.


  torch.__version__  = 2.0.1+cu117


  error: subprocess-exited-with-error

  × python setup.py egg_info did not run successfully.
  │ exit code: 1
  ╰─> See above for output.
  cwd: /tmp/pip-install-wub410gf/flash-attn_06d2dd1c6ec841859f0d8ad4972cc418/
  Preparing metadata (setup.py) ... error
error: metadata-generation-failed

× Encountered error while generating package metadata.
╰─> See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

**Comment by tridao** [comment]
Can you try running with `pip install flash-attn -vv`? What's the error message there?

**Comment by tmm1** [comment]
i can reproduce this as follows, on a machine where `nvcc` is not available:

```
wget https://files.pythonhosted.org/packages/76/8d/354f4691676267243587bb8f672ca85ee3a7877fdb252ca700be533f07dd/flash_attn-2.0.6.tar.gz

tar zxvf flash_attn-2.0.6

cd flash_attn-2.0.6

python3 setup.py bdist_wheel
```

which produces

```
fatal: not a git repository (or any of the parent directories): .git


torch.__version__  = 2.0.1+cu117


Traceback (most recent call last):
  File "/tmp/flash_attn-2.0.6/setup.py", line 140, in <module>
    raise_if_cuda_home_none("flash_attn")
  File "/tmp/flash_attn-2.0.6/setup.py", line 86, in raise_if_cuda_home_none
    raise RuntimeError(
RuntimeError: flash_attn was requested, but nvcc was not found.  Are you sure your environment has nvcc available?  If you're installing within a container from https://hub.docker.com/r/pytorch/pytorch, only images whose names contain 'devel' will provide nvcc.
```

cc @teknium1

**Comment by tmm1** [comment]
> basically we never get to `CachedWheelCommand` because the body of setup.py raises an error first.

nope, my mistake. I missed the call to `raise_if_cuda_home_none` inside `CachedWheelCommand`

**Comment by tmm1** [comment]
workaround:

>CUDA_HOME=none FLASH_ATTENTION_SKIP_CUDA_BUILD=TRUE python setup.py bdist_wheel

**Comment by surak** [comment]
this is not fixed as of july 23, 2024 with flash_attn 2.6.1
